swiftnav/src/coords/mod.rs

```diff
@@ -0,0 +1,534 @@
+//!
+//! Where the 'radius of curvature', $N(\phi)$, is defined as:
+//!
+//! $$N(\phi) = \frac{a}{\sqrt{1-e^2\sin^2 \phi}}$$
+//!
+//! and $a$ is the WGS84 semi-major axis and $e$ is the WGS84
+//! eccentricity.
+//!
+//! # Cartesian to Geodetic
+//!
+//! Conversion from Cartesian to geodetic coordinates is a much harder problem
+//! than conversion from geodetic to Cartesian. There is no satisfactory closed
+//! form solution but many different iterative approaches exist.
+//!
+//! Here we implement a relatively new algorithm due to Fukushima (2006) that is
+//! very computationally efficient, not requiring any transcendental function
+//! calls during iteration and very few divisions. It also exhibits cubic
+//! convergence rates compared to the quadratic rate of convergence seen with
+//! the more common algortihms based on the Newton-Raphson method.
+//!
+//! ## References
+//! * "A comparison of methods used in rectangular to Geodetic Coordinates
+//!   Transformations", Burtch R. R. (2006), American Congress for Surveying
+//!   and Mapping Annual Conference. Orlando, Florida.
+//! * "Transformation from Cartesian to Geodetic Coordinates Accelerated by
+//!   Halley’s Method", T. Fukushima (2006), Journal of Geodesy.
+//!
+//! # Examples
+//!
+//! ```rust
+//! # use swiftnav::coords::{ECEF, LLHDegrees};
+//! # use float_eq::assert_float_eq;
+//! let position_llh = LLHDegrees::new(37.791, -122.395, 0.0);
+//! let position_ecef = position_llh.to_ecef();
+//!
+//! // Approximate velocity of the local crust in meters/year
+//! let velocity_vector = ECEF::new(-0.02, 0.03, 0.01);
+//! let velocity_ned = velocity_vector.ned_vector_at(&position_ecef);
+//! assert_float_eq!(velocity_ned.n(), 0.02, abs <= 0.01);
+//! assert_float_eq!(velocity_ned.e(), -0.03, abs <= 0.01);
+//! assert_float_eq!(velocity_ned.d(), 0.0, abs <= 0.01);
+//!
+//! let years = 3.5;
+//! let future_ecef = position_ecef + (years * velocity_vector);
+//!
+//! let sutro_tower = LLHDegrees::new(37.75523, -122.45284, 254.2);
+//! let bearing = position_ecef.azel_of(&sutro_tower.into());
+//! assert_float_eq!(bearing.az(), 4.05, abs <= 0.01);
+//! assert_float_eq!(bearing.el(), 0.04, abs <= 0.01);
+//! ```
+
+mod ecef;
+mod ellipsoid;
+mod llh;
+mod ned;
+
+pub use ecef::*;
+pub use ellipsoid::*;
+pub use llh::*;
+pub use ned::*;
+
+use crate::{
+    reference_frame::{get_transformation, ReferenceFrame, TransformationNotFound},
+    time::GpsTime,
+};
+use nalgebra::{ArrayStorage, Vector2};
+
+/// WGS84 local horizontal coordinates consisting of an Azimuth and Elevation, with angles stored as radians
+///
+/// Azimuth can range from $0$ to $2\pi$. North has an azimuth of $0$, east has an azimuth of $\frac{\pi}{2}$
+///
+/// Elevation can range from $-\frac{\pi}{2}$ to $\frac{\pi}{2}$. Up has an elevation of $\frac{\pi}{2}$, down an elevation of $-\frac{\pi}{2}$
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd, Default)]
+pub struct AzimuthElevation(Vector2<f64>);
```

Comment:
I'll stop leaving comments about the `From` impls from here on, but comment here haha